### PR TITLE
enh(CPlusPlus): Binary Search Tree and Iterative Traversal

### PR DESCRIPTION
**Pull request by shinchancode**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] New algorithm
- [ ] Optimization in previous algorithms
- [ ] Code style update
- [ ] Refactor
- [ ] Documentation
- [x] Other, please describe:

**Briefly describe the changes in this PR**
1. I have added iterative method also for traversal of tree using stak for better understanding.
2. I have added three functions in Binary search tree : finding maximum, minimum value in a BST and printing all lead nodes present in the tree.

**Other information:**

### CODE REVIEW

File: algorithms/CPlusPlus/Trees/binary-search-tree.cpp [hunk 1]
```diff
@@ -158,83 +159,155 @@ bool isperfect(TreeNode* root){
...
+   } 
+
+   if (root->left) 
+       leaf_nodes(root->left); 
+
+   if (root->right) 
+       leaf_nodes(root->right); 
+} 
+
+void f_min(TreeNode* root) //Find Minimum element from root
```
Comment (Arsenic-ATG) [inline algorithms/CPlusPlus/Trees/binary-search-tree.cpp:187]: **Suggestion :**
in are general used case, one would rather use this function to get the minimum value in the tree and use it somewhere else. 

So isn't it better for this function ( and `f_max ()` ) to return the minimum value to the caller and let the caller use it for whatever it want the minimum value for ?

Comment (shinchancode) [reply algorithms/CPlusPlus/Trees/binary-search-tree.cpp:187]: Ok sir, I will work on it. I will try to add delete function. There we need minimum value.
Thank you for suggestion

File: algorithms/CPlusPlus/Trees/binary-search-tree.cpp [hunk 2]
```diff
@@ -158,83 +159,226 @@ bool isperfect(TreeNode* root){
...
+
+    if(isperfect(root)) { std::cout << "The binary tree is PERFECT" << std::endl; }
+    else           { std::cout << "The binary tree is not PERFECT" << std::endl; }
+
+    std::cout << "Leaf Nodes present in the binary tree are : ";
+    leaf_nodes(root);
+    std::cout<< std::endl;
+    
+    n = f_max(root);
+    if(f_max(root)!=NULL)
```
Comment (Arsenic-ATG) [inline algorithms/CPlusPlus/Trees/binary-search-tree.cpp:333]: We already have `n` holding the result of `f_max(root)`, so this function call overhead can be avoided.

```suggestion
    if(n)
```

File: algorithms/CPlusPlus/Trees/binary-search-tree.cpp [hunk 3]
```diff
@@ -158,83 +159,226 @@ bool isperfect(TreeNode* root){
...
+    {
+        std::cout<<"Maximum Value Present in the tree is : "<<n->data<< std::endl;
+    }
+
+    n=f_min(root);
+    if(f_max(root)!=NULL)
```
Comment (Arsenic-ATG) [inline algorithms/CPlusPlus/Trees/binary-search-tree.cpp:339]: Same here
```suggestion
    if(n)
```

File: algorithms/CPlusPlus/Trees/binary-search-tree.cpp (hunk 1 above)
Comment (Arsenic-ATG) [reply algorithms/CPlusPlus/Trees/binary-search-tree.cpp]: It's nice that you also added a used case in the program also.

**Review by Arsenic-ATG — APPROVED** [review]

**Pull request merged**